README.md
# Firebase Token Generator - PHP

[Firebase Custom Login](https://www.firebase.com/docs/web/guide/simple-login/custom.html)
gives you complete control over user authentication by allowing you to authenticate users
with secure JSON Web Tokens (JWTs). The auth payload stored in those tokens is available
for use in your Firebase [security rules](https://www.firebase.com/docs/security/api/rule/).
This is a token generator library for PHP which allows you to easily create those JWTs.


## Dependencies

The Firebase PHP token generator library depends on [PHP-JWT](https://github.com/firebase/php-jwt).


## A Note About Security

**IMPORTANT:** Because token generation requires your Firebase Secret, you should only generate
tokens on *trusted servers*. Never embed your Firebase Secret directly into your application and
never share your Firebase Secret with a connected client.


## Generating Tokens

To generate tokens, you'll need your Firebase Secret which you can find by entering your Firebase
URL into a browser and clicking the "Secrets" tab on the left-hand navigation menu.

Once you've downloaded the library and grabbed your Firebase Secret, you can generate a token with
this snippet of PHP code:

```php
<?php
  include_once "FirebaseToken.php";

  $tokenGen = new Services_FirebaseTokenGenerator("<YOUR_FIREBASE_SECRET>");
  $token = $tokenGen->createToken(array("uid" => "exampleID"));
?>
```

The payload passed to `createToken()` is made available for use within your
security rules via the [`auth` variable](https://www.firebase.com/docs/security/api/rule/auth.html).
This is how you pass trusted authentication details (e.g. the client's user ID)
to your Firebase security rules. The payload can contain any data of your
choosing, however it must contain a "uid" key, which must be a string of less
than 256 characters. The generated token must be less than 1024 characters in
total.


## Token Options

A second `options` argument can be passed to `createToken()` to modify how Firebase treats the
token. Available options are:

* **expires** (number or DateTime) - A timestamp (as number of seconds since the epoch) or a `DateTime`
denoting the time after which this token should no longer be valid.

* **notBefore** (number or DateTime) - A timestamp (as number of seconds since the epoch) or a `DateTime`
denoting the time before which this token should be rejected by the server.

* **admin** (boolean) - Set to `True` if you want to disable all security rules for this client.
This will provide the client with read and write access to your entire Firebase database.

* **debug** (boolean) - Set to `True` to enable debug output from your security rules. You should
generally *not* leave this set to `True` in production (as it slows down the rules implementation
and gives your users visibility into your rules), but it can be helpful for debugging.

Here is an example of how to use the second `options` argument:

```php
<?php
  include_once "FirebaseToken.php";

  $tokenGen = new Services_FirebaseTokenGenerator("<YOUR_FIREBASE_SECRET>");
  $token = $tokenGen->createToken(array("uid" => "exampleID"), array("admin" => True));
?>
```


CHANGELOG.md
# CHANGELOG

## Unreleased

- The required PHP version is now >= 5.4
- Changed usage, please see the README for documentation.

## 2.0.1 - 2015-04-02

Specifying the PHP-JWT version more specifically

## 2.0.0 - 2014-09-15

Additional validation to ensure tokens contain a "uid" field unless they have the "admin" option set to `true`.

## 1.0.0 - 2014-09-04

Initial release


tests/FirebaseTokenTest.php
<?php

include_once "FirebaseToken.php";

class FirebaseTokenTest extends PHPUnit_Framework_TestCase {
  function testCreate() {
    $key = "0014ae3b1ded44de9d9f6fc60dfd1c64";
    $tokenGen = new Services_FirebaseTokenGenerator($key);
    $token = $tokenGen->createToken(array("foo" => "bar", "baz" => "boo", "uid" => "blah"));

    $data = JWT::decode($token, $key, array('HS256'));
    $this->assertEquals("bar", $data->d->foo);
    $this->assertEquals("boo", $data->d->baz);
    $this->assertInternalType("integer", $data->iat);
  }

  function testAdminDebug() {
    $key = "foobar";
    $tokenGen = new Services_FirebaseTokenGenerator($key);
    $token = $tokenGen->createToken(null, array("admin" => true, "debug" => true));

    $data = JWT::decode($token, $key, array('HS256'));
    $this->assertTrue($data->admin);
    $this->assertTrue($data->debug);
  }

  function testMalformedKeyThrowsException() {
    $this->setExpectedException("UnexpectedValueException", 'Invalid secret provided');
    $tokenGen = new Services_FirebaseTokenGenerator(1234567890);
  }

  function testExpires() {
    $key = "barfoo";
    $tokenGen = new Services_FirebaseTokenGenerator($key);
    $expires = time() + 1000;
    $token = $tokenGen->createToken(array("uid" => "blah"), array("expires" => $expires));

    $data = JWT::decode($token, $key, array('HS256'));
    $this->assertEquals($expires, $data->exp);
  }

  function testNotBeforeObject() {
    $key = "barfoo";
    $tokenGen = new Services_FirebaseTokenGenerator($key);
    $notBefore = new DateTime("now", new DateTimeZone('America/Los_Angeles'));
    $token = $tokenGen->createToken(array("uid" => "blah"), array("notBefore" => $notBefore));

    $data = JWT::decode($token, $key, array('HS256'));
    $this->assertEquals($notBefore->getTimestamp(), $data->nbf);
  }

  function testNoUID() {
    $key = "barfoo";
    $tokenGen = new Services_FirebaseTokenGenerator($key);
    $this->setExpectedException("Exception", 'Services_FirebaseTokenGenerator->createToken: data must contain a "uid" key that must be a string.');
    $token = $tokenGen->createToken(array("blah" => 5));
  }

  function testInvalidUID() {
    $key = "barfoo";
    $tokenGen = new Services_FirebaseTokenGenerator($key);
    $this->setExpectedException("Exception", 'Services_FirebaseTokenGenerator->createToken: data must contain a "uid" key that must be a string.');
    $token = $tokenGen->createToken(array("uid" => 5, "blah" => 5));
  }

  function testUIDMaxLength() {
    $key = "barfoo";
    $tokenGen = new Services_FirebaseTokenGenerator($key);
    //length:                                               10        20        30        40        50        60        70        80        90       100       110       120       130       140       150       160       170       180       190       200       210       220       230       240       250   256
    $token = $tokenGen->createToken(array("uid" => "1234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456"));
  }

  function testUIDTooLong() {
    $key = "barfoo";
    $tokenGen = new Services_FirebaseTokenGenerator($key);
    $this->setExpectedException("Exception", 'Services_FirebaseTokenGenerator->createToken: data must contain a "uid" key that must not be longer than 256 bytes.');
    //length:                                               10        20        30        40        50        60        70        80        90       100       110       120       130       140       150       160       170       180       190       200       210       220       230       240       250    257
    $token = $tokenGen->createToken(array("uid" => "12345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567"));
  }

  function testUIDMinLength() {
    $key = "barfoo";
    $tokenGen = new Services_FirebaseTokenGenerator($key);
    $token = $tokenGen->createToken(array("uid" => ""));
  }

  function testTokenTooLong() {
    $key = "barfoo";
    $tokenGen = new Services_FirebaseTokenGenerator($key);
    $this->setExpectedException("Exception", "Services_FirebaseTokenGenerator->createToken: generated token is too large.  Token cannot be larger than 1024 bytes.");
    $token = $tokenGen->createToken(array("uid" => "blah", "longVar" => "123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345612345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234561234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456"));
  }

  function testNoUIDWithAdmin() {
    $key = "barfoo";
    $tokenGen = new Services_FirebaseTokenGenerator($key);
    $token = $tokenGen->createToken(null, array("admin" => true));
    $token = $tokenGen->createToken(array(), array("admin" => true));
    $token = $tokenGen->createToken(array("foo" => "bar"), array("admin" => true));
  }

  function testInvalidUIDWithAdmin1() {
    $key = "barfoo";
    $tokenGen = new Services_FirebaseTokenGenerator($key);
    $this->setExpectedException("Exception", 'Services_FirebaseTokenGenerator->createToken: data must contain a "uid" key that must be a string.');
    $token = $tokenGen->createToken(array("uid" => 1), array("admin" => true));
  }

  function testInvalidUIDWithAdmin2() {
    $key = "barfoo";
    $tokenGen = new Services_FirebaseTokenGenerator($key);
    $this->setExpectedException("Exception", 'Services_FirebaseTokenGenerator->createToken: data must contain a "uid" key that must be a string.');
    $token = $tokenGen->createToken(array("uid" => null), array("admin" => true));
  }

  function testInvalidUIDWithAdmin3() {
    $key = "barfoo";
    $tokenGen = new Services_FirebaseTokenGenerator($key);
    $this->setExpectedException("Exception", "Services_FirebaseTokenGenerator->createToken: data must be null or an associative array of token data.");
    $token = $tokenGen->createToken("foo", array("admin" => true));
  }

  function testEmptyDataAndNoOptionsThrowsException() {
    $key = "barfoo";
    $tokenGen = new Services_FirebaseTokenGenerator($key);
    $this->setExpectedException("Exception", "Services_FirebaseTokenGenerator->createToken: data is empty and no options are set.  This token will have no effect on Firebase.");
    $tokenGen->createToken(null);
  }

  function testMalformedDataThrowsException() {
    $key = "barfoo";
    $tokenGen = new Services_FirebaseTokenGenerator($key);
    $this->setExpectedException("UnexpectedValueException", "Malformed UTF-8 characters, possibly incorrectly encoded");
    $tokenGen->createToken("\xB1\x31");
  }
}

?>


tests/TokenGeneratorTest.php
<?php

/*
 * This file is part of the Firebase Token Generator.
 *
 * This source file is subject to the license that is bundled
 * with this source code in the file LICENSE.
 */

namespace Firebase\Token\Test;

use Firebase\Token\TokenGenerator;

class TokenGeneratorTest extends \PHPUnit_Framework_TestCase
{
    /**
     * @var \Firebase\Token\TokenGenerator
     */
    protected $generator;

    /**
     * @var string
     */
    protected $secret;

    protected function setUp()
    {
        $this->secret    = 'secret';
        $this->generator = new TokenGenerator($this->secret);
    }

    public function testDefaultOptions()
    {
        $token = $this->generator->setData(['uid' => 'uid'])->create();
        $data  = $this->decodeToken($token);

        $this->assertFalse($data->admin);
        $this->assertFalse($data->debug);
        $this->assertObjectNotHasAttribute('exp', $data);
        $this->assertObjectNotHasAttribute('nbf', $data);
    }

    public function testSetOptions()
    {
        $notBefore = time() - 10;
        $expires   = time() + 10;

        $token = $this->generator
            ->setData(['uid' => 'uid'])
            ->setOptions([
                'admin'     => true,
                'debug'     => true,
                'expires'   => $expires,
                'notBefore' => $notBefore,
            ])
            ->create();
        $check = $this->decodeToken($token);

        $this->assertTrue($check->admin);
        $this->assertTrue($check->debug);
        $this->assertEquals($expires, $check->exp);
        $this->assertEquals($notBefore, $check->nbf);
    }

    public function testTriggerExceptionOnUnsupportedOption()
    {
        $this->setExpectedExceptionRegExp('\Firebase\Token\TokenException', '/^Unsupported option "foo"/');
        $this->generator->setOption('foo', 'bar');
    }

    /**
     * @param string $name
     * @param string $expectedType
     * @param mixed  $invalidValue
     *
     * @dataProvider invalidOptionsProvider
     */
    public function testTriggerExceptionOnInvalidOption($name, $expectedType, $invalidValue)
    {
        $this->setExpectedException(
            '\Firebase\Token\TokenException',
            sprintf('Invalid option "%s". Expected %s, but %s given', $name, $expectedType, gettype($invalidValue))
        );

        $this->generator->setOption($name, $invalidValue);
    }

    public function invalidOptionsProvider()
    {
        return [
            ['admin', 'bool', 'foo'],
            ['debug', 'bool', 'foo'],
            ['expires', 'int or DateTime', 'foo'],
            ['notBefore', 'int or DateTime', 'foo'],
        ];
    }

    /**
     * @expectedException \Firebase\Token\TokenException
     * @expectedExceptionMessage The Firebase secret must be a string, integer given
     */
    public function testTriggerExceptionOnInvalidSecret()
    {
        new TokenGenerator(1);
    }

    /**
     * @expectedException \Firebase\Token\TokenException
     * @expectedExceptionMessage No uid provided in data and admin option not set.
     */
    public function testTriggerExceptionOnMissingUid()
    {
        $this->generator->create();
    }

    /**
     * @expectedException \Firebase\Token\TokenException
     * @expectedExceptionMessage The uid must be a string, integer given
     */
    public function testTriggerExceptionOnNonStringUid()
    {
        $this->generator->setData(['uid' => 1])->create();
    }

    public function testTriggerExceptionIfUidIsTooLong()
    {
        $this->setExpectedException(
            '\Firebase\Token\TokenException',
            sprintf(
                'The provided uid is longer than %d bytes (%d)',
                TokenGenerator::MAX_UID_SIZE, TokenGenerator::MAX_UID_SIZE + 1
            )
        );

        $this->generator->setData(['uid' => str_repeat('x', TokenGenerator::MAX_UID_SIZE + 1)])->create();
    }

    /**
     * @expectedException \Firebase\Token\TokenException
     * @expectedExceptionMessage The provided uid is empty
     */
    public function testTriggerExceptionIfUidIsEmpty()
    {
        $this->generator->setData(['uid' => ''])->create();
    }

    public function testTriggerExceptionIfGeneratedTokenIsTooLong()
    {
        $expectedMessagePattern = sprintf(
            '/^The generated token is larger than %d bytes \(\d+\)$/', TokenGenerator::MAX_TOKEN_SIZE
        );

        $this->setExpectedExceptionRegExp('\Exception', $expectedMessagePattern);

        $this->generator
            ->setData(['uid' => 'uid', 'foo' => str_repeat('x', TokenGenerator::MAX_TOKEN_SIZE + 1)])
            ->create();
    }

    public function testAllowEmptyUidIfAdminOptionIsSet()
    {
        $token = $this->generator->setOption('admin', true)->create();
        $data  = $this->decodeToken($token);

        $this->assertTrue($data->admin);
    }

    /**
     * @expectedException \Firebase\Token\TokenException
     */
    public function testTriggerExceptionOnJwtException()
    {
        $this->generator
            ->setData(['uid' => "\xB1\x31"]) // Invalid UTF-8 character
            ->create();
    }

    /**
     * Decodes a token.
     *
     * @param string $token
     *
     * @return object
     */
    protected function decodeToken($token)
    {
        return \JWT::decode($token, $this->secret, ['HS256']);
    }
}


src/TokenGenerator.php
<?php

/*
 * This file is part of the Firebase Token Generator.
 *
 * This source file is subject to the license that is bundled
 * with this source code in the file LICENSE.
 */

namespace Firebase\Token;

/**
 * Creates JWT token to to authenticate requests to a Firebase application.
 */
class TokenGenerator
{
    // Sizes in bytes
    const MAX_UID_SIZE   = 256;
    const MAX_TOKEN_SIZE = 1024;

    /**
     * The token data.
     *
     * @var mixed[]
     */
    private $data;

    /**
     * The token options.
     *
     * @var mixed[]
     */
    private $options;

    /**
     * Initializes the generator.
     *
     * @param string $secret The Firebase secret.
     *
     * @throws TokenException If an invalid secret has been given.
     */
    public function __construct($secret)
    {
        if (!is_string($secret)) {
            throw new TokenException(
                sprintf('The Firebase secret must be a string, %s given', gettype($secret))
            );
        }

        $this->secret  = $secret;
        $this->version = 0;
        $this->data    = [];

        // Default options
        $this->options = [
            'admin'     => false,
            'debug'     => false,
            'expires'   => null,
            'notBefore' => null,
        ];
    }

    /**
     * Sets the token data.
     *
     * @param array $data An array of data you wish to associate with the token.
     *                    It will be available as the variable "auth" in the Firebase rules engine.
     *
     * @return static
     */
    public function setData(array $data)
    {
        $this->data = $data;

        return $this;
    }

    /**
     * Sets multiple options.
     *
     * @see setOption()
     *
     * @param array $options The options.
     *
     * @throws TokenException If an invalid option has been given.
     *
     * @return static
     */
    public function setOptions(array $options)
    {
        foreach ($options as $name => $value) {
            $this->setOption($name, $value);
        }

        return $this;
    }

    /**
     * Sets an option.
     *
     * @param string $name  The option name.
     * @param mixed  $value The option value.
     *
     * @throws TokenException If an invalid option has been given.
     *
     * @return static
     */
    public function setOption($name, $value)
    {
        if (!array_key_exists($name, $this->options)) {
            throw new TokenException(
                sprintf(
                    'Unsupported option "%s". Valid options are: ', $name, implode(', ', array_keys($this->options))
                )
            );
        }

        switch ($name) {
            case 'admin':
            case 'debug':
                if (!is_bool($value)) {
                    throw new TokenException(
                        sprintf('Invalid option "%s". Expected %s, but %s given', $name, 'bool', gettype($value))
                    );
                }
                break;
            case 'expires':
            case 'notBefore':
                if (!is_int($value) && !($value instanceof \DateTime)) {
                    throw new TokenException(
                        sprintf(
                            'Invalid option "%s". Expected %s, but %s given',
                            $name, 'int or DateTime', gettype($value)
                        )
                    );
                }

                if (is_int($value)) {
                    $value = \DateTime::createFromFormat('U', $value);
                }
                break;
        }

        $this->options[$name] = $value;

        return $this;
    }

    /**
     * Creates the token.
     *
     * @throws TokenException If the token couldn't be generated or is invalid.
     *
     * @return string The JWT token.
     */
    public function create()
    {
        $this->validate();

        $claims        = $this->processOptions();
        $claims['d']   = $this->data;
        $claims['v']   = 0;
        $claims['ait'] = time();

        try {
            $token = \JWT::encode($claims, $this->secret, 'HS256');
        } catch (\Exception $e) {
            throw new TokenException($e->getMessage(), null, $e);
        }

        if (($tokenSize = mb_strlen($token, '8bit')) > static::MAX_TOKEN_SIZE) {
            throw new TokenException(
                sprintf('The generated token is larger than %d bytes (%d)', static::MAX_TOKEN_SIZE, $tokenSize)
            );
        }

        return $token;
    }

    /**
     * Parses provided options into a claims array.
     *
     * @return array The claims.
     */
    private function processOptions()
    {
        $claims = [];

        foreach ($this->options as $name => $value) {
            switch ($name) {
                case 'expires':
                    if ($value instanceof \DateTime) {
                        $claims['exp'] = $value->getTimestamp();
                    }
                    break;
                case 'notBefore':
                    if ($value instanceof \DateTime) {
                        $claims['nbf'] = $value->getTimestamp();
                    }
                    break;
                default:
                    $claims[$name] = $value;
                    break;
            }
        }

        return $claims;
    }

    /**
     * Validates the combination of data and options.
     *
     * @throws TokenException If the combination of data and options is invalid.
     */
    private function validate()
    {
        if (false === $this->options['admin']) {
            if (!array_key_exists('uid', $this->data)) {
                throw new TokenException('No uid provided in data and admin option not set.');
            }

            $this->validateUid($this->data['uid']);
        }
    }

    /**
     * Validates an uid.
     *
     * @param string $uid The uid.
     *
     * @throws TokenException If the uid is invalid.
     */
    private function validateUid($uid)
    {
        if (!is_string($uid)) {
            throw new TokenException(sprintf('The uid must be a string, %s given', gettype($uid)));
        }

        $uidSize = mb_strlen($uid, '8bit');

        if ($uidSize > static::MAX_UID_SIZE) {
            throw new TokenException(
                sprintf('The provided uid is longer than %d bytes (%d)', static::MAX_UID_SIZE, $uidSize)
            );
        }

        if (0 === $uidSize) {
            throw new TokenException('The provided uid is empty');
        }
    }
}


FirebaseToken.php
<?php

$jwt_filename = dirname(__FILE__) . '/JWT.php';

if (!file_exists($jwt_filename)) {
    // try Composer autoload
    $jwt_filename = dirname(__FILE__) . '/vendor/autoload.php';
}

if (file_exists($jwt_filename)) {
    include_once $jwt_filename;
}

class Services_FirebaseTokenGenerator
{
    private $version = 0;

    /**
     * Example usage:
     *     $tokenGen = new Services_FirebaseTokenGenerator("0014ae3b1ded44de9d9f6fc60dfd1c64");
     *     $tokenGen->createToken(array("id" => "foo", "bar" => "baz"));
     *
     * @access  public
     * @param   string       $secret   The API secret for the Firebase you
     *                                 want to generate an auth token for.
     */
    public function __construct($secret)
    {
        if (!is_string($secret)) {
            throw new UnexpectedValueException("Invalid secret provided");
        }
        $this->secret = $secret;
    }

    /**
     * @access  public
     * @param   array|object $data     An object or array of data you wish
     *                                 to associate with the token. It will
     *                                 be available as the variable "auth" in
     *                                 the Firebase rules engine.
     * @param   object       $options  Optional. An associative array with
     *                                 the developer supplied options for this
     *                                 token. The following keys are recognized:
     *
     *                                   'admin': Set to true if you want this
     *                                   token to bypass all security rules.
     *                                   Defaults to false.
     *
     *                                   'debug': Set to true if you want to
     *                                   enable debug output from your security
     *                                   rules.
     *
     *                                   'expires': Set to a number (seconds
     *                                   since epoch) or a DateTime object that
     *                                   specifies the time at which the token
     *                                   should expire.
     *
     *                                   'notBefore': Set to a number (seconds
     *                                   since epoch) or a DateTime object that
     *                                   specifies the time before which the
     *                                   should be rejected by the server.
     *
     *
     * @return  string       A Firebase auth token.
     */
    public function createToken($data, $options = null)
    {
        $funcName = 'Services_FirebaseTokenGenerator->createToken';

        // If $data is JSONifiable, let it pass.
        $json = json_encode($data);
        if (function_exists("json_last_error") && $errno = json_last_error()) {
            $this->handleJSONError($errno);
        } else if ($json === "null" && $data !== null) {
            throw new UnexpectedValueException("Data is not valid JSON");
        } else if (empty($data) && empty($options)) {
            throw new Exception($funcName . ": data is empty and no options are set.  This token will have no effect on Firebase.");
        }

        $claims = array();
        if (is_array($options)) {
            $claims = $this->_processOptions($options);
        }

        $this->_validateData($funcName, $data, (isset($claims['admin']) && ($claims["admin"] == true)));

        $claims["d"] = $data;
        $claims["v"] = $this->version;
        $claims["iat"] = time();

        $token = JWT::encode($claims, $this->secret, "HS256");
        if (strlen($token) > 1024) {
            throw new Exception($funcName . ": generated token is too large.  Token cannot be larger than 1024 bytes.");
        }
        return $token;
    }

    /**
     * Parses provided options into a claims object.
     *
     * @param object $options Options as passed by the developer to createToken.
     *
     * @return array A claims array in which the options are stored.
     */
    private static function _processOptions($options) {
        $claims = array();
        foreach ($options as $key => $value) {
            switch ($key) {
                case "admin":
                    $claims["admin"] = $value;
                    break;
                case "debug":
                    $claims["debug"] = $value;
                    break;
                case "expires":
                case "notBefore":
                    $code = "exp";
                    if ($key == "notBefore") {
                        $code = "nbf";
                    }
                    switch (gettype($value)) {
                        case "integer":
                            $claims[$code] = $value;
                            break;
                        case "object":
                            if ($value instanceof DateTime) {
                                $claims[$code] = $value->getTimestamp();
                            } else {
                                throw new UnexpectedValueException(
                                    "Provided " . $key .
                                    " option is not a DateTime object");
                            }
                            break;
                        default:
                            throw new UnexpectedValueException(
                                "Provided " . $key .
                                " option is invalid " . $value);
                    }
                    break;
                default:
                    throw new UnexpectedValueException(
                        "Invalid key " . $key . " provided in options");
            }
        }
        return $claims;
    }

    /**
     * Validates provided data object, throwing Exceptions where necessary.
     *
     * @param string $funcName the function name string for error message reporting.
     * @param array $data the token data to be validated.
     * @param boolean $isAdminToken whether the admin flag has been set.
     */
     private static function _validateData($funcName, $data, $isAdminToken) {
        if (!is_null($data) && !is_array($data)) {
            throw new Exception($funcName . ": data must be null or an associative array of token data.");
        }
        $containsUID = (is_array($data) && array_key_exists("uid", $data));
        if ((!$containsUID && !$isAdminToken) || ($containsUID && !is_string($data["uid"]))) {
            throw new Exception($funcName . ": data must contain a \"uid\" key that must be a string.");
        } else if ($containsUID && (strlen($data["uid"]) > 256)) {
            throw new Exception($funcName . ": data must contain a \"uid\" key that must not be longer than 256 bytes.");
        }
    }

    /**
     * @access  private
     * @param   int          $errno    An error number from json_last_error()
     *
     * @return  void
     */
    private static function handleJSONError($errno)
    {
        $messages = array(
            JSON_ERROR_DEPTH => 'Maximum stack depth exceeded',
            JSON_ERROR_CTRL_CHAR => 'Unexpected control character found',
            JSON_ERROR_SYNTAX => 'Syntax error, malformed JSON',
            JSON_ERROR_UTF8 => 'Malformed UTF-8 characters, possibly incorrectly encoded'
        );
        throw new UnexpectedValueException(isset($messages[$errno])
            ? $messages[$errno]
            : 'Unknown JSON error: ' . $errno
        );
    }
}
